File: sort.py
def merge(arr, l, m, r):
    n1 = m - l + 1
    n2 = r - m
    
    L = [0] * n1
    R = [0] * n2 
    
    for i in range(n1):
        L[i] = arr[l+i]
    for j in range(n2):
        R[j] = arr[m+1+j]
    
    i = 0
    j = 0
    k = l
    
    while i < n1 and j < n2:
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1
    
    while i < n1:
        arr[k] = L[i]
        i += 1
        k += 1
    
    while j < n2:
        arr[k] = R[j]
        j += 1
        k += 1
        
def merge_sort(arr, l, r):
    if l < r:
        m = l + (r-l)/2
        merge_sort(arr, l, m)
        merge_sort(arr, m+1, r)
        merge(arr, l, m, r)
 

test = [6,2,8,4,2]
print test

merge_sort(test, 0, len(test)-1)
print test

File: sort.php
<?php

class Step
{
    public $value;

    public function __construct($value)
    {
        $this->value = $value;
    }
}

class Sorting
{
    public static function mergeSort(&$input)
    {
    }

    public static function selectionSort(&$input)
    {

    }

    public static function insertionSort(&$input)
    {
        $steps = array();
        $n = count($input);
        for ($i=1; $i<$n; $i++) {
            $key = $input[$i];
            $j = $i-1;

            while ($j>=0 && $input[$j] > $key) {
                $input[$j+1] = $input[$j];
                $j = $j-1;
                array_push($steps, new Step($input));
            }

            $input[$j+1] = $key;
        }
        return $steps;
    }

    public static function quickSort(&$input)
    {

    }

    private static function swap(&$a, &$b)
    {
        $temp = $a;
        $a = $b;
        $b = $temp;
    }
}
function print_array($a)
{
    foreach ($a as $number) {
        printf("$number ");
    }
    printf("\n");
}

$numbers = range(1,20);
shuffle($numbers);
$steps = Sorting::insertionSort($numbers);
var_export($steps);
?>
